### PR TITLE
SUS-1716: Wall permissions checking refactor

### PR DESCRIPTION
**Pull request by mszabo-wikia**
## Rationale
Wall has a custom method for checking permissions which implements instance cache as well. This is redundant, since user-permissions service also has an instance cache for user permissions. Additionally, we are also checking blocked status even when we're just rendering user interface components, which triggers several redundant Phalanx lookups.

## Changes
Only user permissions are checked when deciding whether to render user interface components. Blocked status is only checked if the user actually tries to perform an action (e.g. posting a message, or deleting a message etc.). If they are blocked, they are shown an informative error modal with their block details.

https://wikia-inc.atlassian.net/browse/SUS-1716

### CODE REVIEW

File: includes/wikia/nirvana/WikiaDispatcher.class.php [hunk 1]
```diff
@@ -214,7 +214,7 @@ public function dispatch( WikiaApp $app, WikiaRequest $request ) {
 					$result = $controller->$method( $params );
 				}
 
-				if($result === false) {
+				if ( $result === false && $response->getFormat() !== WikiaResponse::FORMAT_JSON ) {
 				   // skip template rendering when false returned
```
Comment (rybmat) [inline includes/wikia/nirvana/WikiaDispatcher.class.php]: should we update this comment as well?

File: extensions/wikia/Wall/WallMessage.class.php [hunk 2]
```diff
@@ -254,86 +246,44 @@ protected function doSaveMetadata( $user, $summary = '', $force = false ) {
...
-			return true;
-		}
-
-		if ( $this->isAuthor( $user ) || $this->isWallOwner( $user ) ) {
-			return !$this->isMarkInProps( WPP_WALL_MODERATORREMOVE );
-		}
-
-		return false;
+	public function canRemove( User $user ) {
+		return
```
Comment (rybmat) [inline extensions/wikia/Wall/WallMessage.class.php]: do we need this newline?

File: extensions/wikia/Wall/WallMessage.class.php [hunk 3]
```diff
@@ -1179,44 +1119,20 @@ public function canReply() {
...
-	}
-
-	public function isVisible( $user ) {
-		if ( !$this->isAdminDelete( $user ) ) {
-			return true;
-		}
-
-		return false;
+	public function canRestore( User $user ) {
+		return
```
Comment (rybmat) [inline extensions/wikia/Wall/WallMessage.class.php]: unnecessary newline IMO

File: extensions/wikia/Wall/WallExternalController.class.php [hunk 4]
```diff
@@ -23,13 +50,52 @@ public function init() {
...
+		$pageTitle = $context->getTitle();
+
+		if ( in_array( $method, static::BLOCKED_USER_PREVENTED_FROM_METHODS ) ) {
+			// check if user is allowed to post on their wall per block options
+			$talkPage = $user->getTalkPage();
+			$isOwnWall = in_array( $method, static::BLOCKED_USER_MAYBE_ALLOWED_ON_OWN_WALL_METHODS ) &&
+				// we may be viewing the Message Wall page, or the Thread page
+				( $pageTitle->equals( $talkPage ) || $pageTitle->getSubjectPage()->equals( $talkPage ) );
+
+			// we need to call getTalkPage twice because NS_USER_WALL is not a talk namespace (wie pan co? i to jest skandal)
```
Comment (rybmat) [inline extensions/wikia/Wall/WallExternalController.class.php:73]: > wie pan co? i to jest skandal

w rzeczy samej to skandal jest!

**Review by rybmat — APPROVED** [review]
few minor remarks, but generally LGTM

File: extensions/wikia/Wall/WallExternalController.class.php (hunk 4 above)
Comment (macbre) [reply extensions/wikia/Wall/WallExternalController.class.php:73]: I granda, panie!

**Review by rybmat — APPROVED** [review]

File: extensions/wikia/Wall/WallExternalController.class.php [hunk 5]
```diff
@@ -11,6 +11,33 @@ class WallExternalController extends WikiaController {
 	use \Wikia\Logger\Loggable;
 
 	/**
+	 * @var string[] BLOCKED_USER_PREVENTED_FROM_METHODS
```
Comment (macbre) [inline extensions/wikia/Wall/WallExternalController.class.php:14]: Really minor thing - do we need this type hints?

File: extensions/wikia/Wall/WallExternalController.class.php [hunk 6]
```diff
@@ -23,13 +50,52 @@ public function init() {
...
+	 * Invoked automatically by Nirvana dispatcher.
 	 *
-	 * Use for external testing of mail template
-	 *  http://www.communitycarenc.org/elements/media/images/under-construction.jpg ;)
+	 * @param User $user user performing action
+	 * @param string $method name of Nirvana method being called
+	 * @return bool whether to prevent the action
 	 */
-	public function mail() {
+	public function preventBlockedUsage( User $user, string $method ) {
```
Comment (macbre) [inline extensions/wikia/Wall/WallExternalController.class.php:62]: :+1:

File: extensions/wikia/Wall/js/Wall.js [hunk 7]
```diff
@@ -547,7 +547,26 @@
...
+							msg.fadeOut('fast', this.proxy(function () {
+								$(data.html).hide().insertBefore(msg).fadeIn('fast');
+							}));
+						} else {
+							msg.fadeOut('fast', function () {
+								msg.remove();
+							});
+						}
+
+						if (typeof (modal) !== 'undefined') {
```
Comment (macbre) [inline extensions/wikia/Wall/js/Wall.js]: No need to wrap `modal` in `()` - `typeof` is an operator

**Review by macbre — APPROVED** [review]
A few minors left above

File: extensions/wikia/Wall/WallExternalController.class.php (hunk 4 above)
Comment (ghost) [reply extensions/wikia/Wall/WallExternalController.class.php:73]: _Skaczą wesoło ochocze parobki, Fałszywym tonem skrzypiciel rzempoli_ :-)

File: extensions/wikia/Wall/WallBaseController.class.php [hunk 8]
```diff
@@ -364,8 +369,8 @@ public function statusInfoBox() {
 				$this->response->setVal( 'statusInfo', $info );
 				$this->response->setVal( 'id', $wallMessage->getId() );
 				if ( $showRemoveOrDeleteInfo ) {
-					$this->response->setVal( 'canRestore', $wallMessage->canRestore( $this->app->wg->User, false ) );
-					$this->response->setVal( 'fastrestore', $wallMessage->canFastRestore( $this->app->wg->User, false ) );
+					$this->response->setVal( 'canRestore', $wallMessage->canRestore( $this->app->wg->User ) );
```
Comment (ghost) [inline extensions/wikia/Wall/WallBaseController.class.php:371]: Does it make sense to use consistent references to the user object? Let's say `$this->user` here and in other methods in this controller?

**Review by ghost — APPROVED** [review]

File: extensions/wikia/Wall/WallBaseController.class.php (hunk 8 above)
Comment (mszabo-wikia) [reply extensions/wikia/Wall/WallBaseController.class.php:371]: We could use the `User` object provided by the available `IContextSource`, as in https://camo.githubusercontent.com/8de516752b8503f0ff8859fa08d8cca80cc9d1c7/687474703a2f2f66617368696f6e61626c796765656b2e636f6d2f77702d636f6e74656e742f75706c6f6164732f323031342f31312f6b697474792d6361742d76696b696e672d6861742d30342e6a7067 . This way we make our code more testable as well by getting rid of global state.

**Pull request closed without merging**